go-controller/pkg/ovn/controller/services/load_balancer.go

```diff
@@ -0,0 +1,456 @@
+				eps:             eps,
+				routerLocalMode: localTrafficPolicy,
+			}
+			perNodeConfigs = append(perNodeConfigs, nodePortLBConfig)
+		}
+
+		// Build up list of vips
+		vips := append([]string{}, service.Spec.ClusterIPs...)
+		// Handle old clusters w/o v6 support
+		if len(vips) == 0 {
+			vips = []string{service.Spec.ClusterIP}
+		}
+		externalVips := []string{}
+		// ExternalIP
+		externalVips = append(externalVips, service.Spec.ExternalIPs...)
+		// LoadBalancer status
+		for _, ingress := range service.Status.LoadBalancer.Ingress {
+			if ingress.IP != "" {
+				externalVips = append(externalVips, ingress.IP)
+			}
+		}
+
+		// if ETP=Local, then treat ExternalIPs and LoadBalancer IPs specially
+		// otherwise, they're just cluster IPs
+		if localTrafficPolicy && len(externalVips) > 0 {
+			externalIPConfig := lbConfig{
+				protocol:        svcPort.Protocol,
+				inport:          svcPort.Port,
+				vips:            externalVips,
+				eps:             eps,
+				routerLocalMode: true,
+			}
+			perNodeConfigs = append(perNodeConfigs, externalIPConfig)
+		} else {
+			vips = append(vips, externalVips...)
+		}
+
+		// Build the clusterIP config
+		clusterIPConfig := lbConfig{
+			protocol:        svcPort.Protocol,
+			inport:          svcPort.Port,
+			vips:            vips,
+			eps:             eps,
+			routerLocalMode: false, // always false for ClusterIPs
+		}
+
+		// Normally, the ClusterIP LB is global (on all node switches and routers),
+		// unless both of the following are true:
+		// - We're in shared gateway mode, and
+		// - Any of the endpoints are host-network
+		//
+		// In that case, we need to create per-node LBs.
+		if globalconfig.Gateway.Mode == globalconfig.GatewayModeShared &&
+			(hasHostEndpoints(eps.V4IPs) || hasHostEndpoints(eps.V6IPs)) {
+			perNodeConfigs = append(perNodeConfigs, clusterIPConfig)
+		} else {
+			clusterConfigs = append(clusterConfigs, clusterIPConfig)
+		}
+	}
+
+	return
+}
+
+// makeLBName creates the load balancer name - used to minimize churn
+func makeLBName(service *v1.Service, proto v1.Protocol, scope string) string {
+	return fmt.Sprintf("Service_%s/%s_%s_%s",
+		service.Namespace, service.Name,
+		proto, scope,
+	)
+}
+
+// buildClusterLBs takes a list of lbConfigs and aggregates them
+// in to one ovn LB per protocol.
+//
+// It takes a list of (proto:[vips]:port -> [endpoints]) configs and re-aggregates
+// them to a list of (proto:[vip:port -> [endpoint:port]])
+// This load balancer is attached to all node switches. In shared-GW mode, it is also on all routers
+func buildClusterLBs(service *v1.Service, configs []lbConfig, nodeInfos []nodeInfo) []ovnlb.LB {
+	nodeSwitches := make([]string, 0, len(nodeInfos))
+	nodeRouters := make([]string, 0, len(nodeInfos))
+	for _, node := range nodeInfos {
+		nodeSwitches = append(nodeSwitches, node.switchName)
+		if globalconfig.Gateway.Mode == globalconfig.GatewayModeShared && node.gatewayRouterName != "" {
```

Comment:
Because the node might not have a GatewayRouter - it could be waiting for creation, or it could have a special annotation that disables gateway router creation for that node.

Added a comment.